vernac/vernacentries.mli

```diff
@@ -44,6 +44,23 @@ val interp_redexp_hook : (Environ.env -> Evd.evar_map -> Genredexpr.raw_red_expr
 (** Helper *)
 val vernac_require_open_proof : pstate:Proof_global.t option -> (pstate:Proof_global.t -> 'a) -> 'a
 
+val with_pstate : pstate:Proof_global.t option -> (pstate:Proof_global.pstate -> 'a) -> 'a
+
+val modify_pstate : pstate:Proof_global.t option -> (pstate:Proof_global.pstate -> Proof_global.pstate) -> Proof_global.t option
+
 (* Flag set when the test-suite is called. Its only effect to display
    verbose information for `Fail` *)
 val test_mode : bool ref
+
+(** For mlg *)
+
+type functional_vernac =
+  | VtDefault of (unit -> unit)
+  | VtModifyProofStack of (pstate:Proof_global.t option -> Proof_global.t option)
+  | VtMaybeOpenProof of (is_nested:bool -> Proof_global.pstate option)
+  | VtOpenProof of (is_nested:bool -> Proof_global.pstate)
+  | VtModifyProof of (pstate:Proof_global.pstate -> Proof_global.pstate)
+  | VtReadProofOpt of (pstate:Proof_global.pstate option -> unit)
+  | VtReadProof of (pstate:Proof_global.pstate -> unit)
```

Comment:
>The main missing bit is that there is no ADT for the typing step (since coqpp just puts there the right adaptor)

Isn't this the kind of ADT you mean?

IMO the main missing bit is that it handles only the proof_global part of the state, but classify is about all of it.

Reply:
Oh, indeed, it is already there; only thing missing then is to make the interpretation of such vernacs internal to vernacentries.

Reply:
Yes, I missed it, sorry.

Reply:
One thing worries me, there is a PR finally removing VtUknown by @maximedenes and here I see VtMaybeOpenProof.
Is it just because the two PR are in parallel and not one ontop of the other right?

Reply:
I'm talking about #9530

Reply:
> Is it just because the two PR are in parallel and not one ontop of the other right?

That's part of it. There's also the funind `Function` command which is classified depending on the arguments. There's also `Functional Scheme` where you get a pstate option but maybe it's always supposed to be None (classified as sideff), and the `Generate graph for` classified as query which I pointed out before.

Reply:
And maybe some in 3rd party stuff, I forgot what I used in overlays.

Reply:
The we should wait for #9530 to be merged, and get rid of the `VtMaybeOpenProof` constructor.

Reply:
What about `Function`?